[WPD]: Enable speculative devirtualizatoin.

hassnaaHamdi opened the pull request:
This patch implements the speculative devirtualization feature in the LLVM backend.
It handles the case of single implementation devirtualization where there is a single
possible callee of a virtual function.
- Add cl::opt 'devirtualize-speculatively' to enable it.
- Flag is disabled by default.
- It works regardless of the visibility of the object.
- Not enabled for LTO for now.

hassnaaHamdi committed:
[WPD]: Enable speculative devirtualizatoin.

- Add cl::opt 'devirtualize-speculatively' to enable it.
- Flag is disabled by default.
- It works regardless of the visibility of the object.

teresajohnson reviewed (commented):
Thanks, a few minor comments but largely looks good.

hassnaaHamdi committed:
resolve review comments:

- Drop commit where we were building ExportSummary as it can't be
  tested in this patch.
- Consider the case where we have public/non-public type tests
- Use LTO flag to differentiate between LTO mode and non-LTO mode.

Change-Id: I52acd6e7352fe1075c9f990aa78dfdc35b89f124

teresajohnson commented on llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp:
```diff
@@ -1325,10 +1367,10 @@ bool DevirtModule::trySingleImplDevirt(
+  // Out of speculative devirtualization mode, if the only implementation has
+  // local linkage, we must promote to external to make it visible to thin LTO
+  // objects.
+  if (!DevirtSpeculatively && TheFn->hasLocalLinkage()) {
```
That's full LTO. In that case no promotion should be needed, as full LTO merges everything into one module. This handling is for ThinLTO when we have hybrid ThinLTO/fullLTO (via LTO unit splitting), and the internal object in the full LTO object must be promoted to be exported to a ThinLTO object (hence the original comment "We can only get here during the ThinLTO export phase"). In that case we always have an ExportSummary (exports the full LTO analysis to the ThinLTO split modules). So I don't think we need to identify that we are in LTO.

hassnaaHamdi committed:
Remove InLTOMode flag.
Add a flag to differentiate between ExportSummary
that is built locally and the external one.

hassnaaHamdi committed:
rebase

teresajohnson reviewed (commented):
Did the building of the ExportSummary get inadvertently removed from this PR? BTW, if you could please avoid force pushing when you update that would make reviewing much easier - as it is I am unable to compare to old versions. See https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes.

hassnaaHamdi committed:
remove HasLocalSummary var as the Summary doesn't get built in this patch

teresajohnson reviewed (approved):
lgtm
